lib/trie/hash.go

```diff
@@ -5,195 +5,285 @@ package trie
 // NodeHeader | Extra partial key length | Partial Key | Value
-func (h *hasher) encode(n node) ([]byte, error) {
+func encodeNode(n node, buffer *bytes.Buffer, parallel bool) (err error) {
 	switch n := n.(type) {
 	case *branch:
-		return h.encodeBranch(n)
+		err := encodeBranch(n, buffer, parallel)
+		if err != nil {
+			return fmt.Errorf("cannot encode branch: %w", err)
+		}
+		return nil
 	case *leaf:
-		return h.encodeLeaf(n)
+		err := encodeLeaf(n, buffer)
+		if err != nil {
+			return fmt.Errorf("cannot encode leaf: %w", err)
+		}
+
+		n.encodingMu.Lock()
+		defer n.encodingMu.Unlock()
+
+		n.encoding = make([]byte, buffer.Len())
+		copy(n.encoding, buffer.Bytes())
+		return nil
 	case nil:
-		return []byte{0}, nil
+		buffer.Write([]byte{0})
+		return nil
+	default:
+		return fmt.Errorf("%w: %T", ErrNodeTypeUnsupported, n)
 	}
-
-	return nil, nil
 }
 
 func encodeAndHash(n node) ([]byte, error) {
-	h := newHasher(false)
-	defer h.returnToPool()
+	buffer := digestBufferPool.Get().(*bytes.Buffer)
+	buffer.Reset()
+	defer digestBufferPool.Put(buffer)
 
-	encChild, err := h.Hash(n)
+	err := hashNode(n, buffer)
 	if err != nil {
 		return nil, err
 	}
 
-	scEncChild, err := scale.Marshal(encChild)
+	scEncChild, err := scale.Marshal(buffer.Bytes())
 	if err != nil {
 		return nil, err
 	}
 	return scEncChild, nil
 }
 
 // encodeBranch encodes a branch with the encoding specified at the top of this package
-func (h *hasher) encodeBranch(b *branch) ([]byte, error) {
+// to the buffer given.
+func encodeBranch(b *branch, buffer *bytes.Buffer, parallel bool) (err error) {
 	if !b.dirty && b.encoding != nil {
-		return b.encoding, nil
+		_, err = buffer.Write(b.encoding)
+		return err
 	}
-	h.tmp.Reset()
 
 	encoding, err := b.header()
-	h.tmp.Write(encoding)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	h.tmp.Write(nibblesToKeyLE(b.key))
-	h.tmp.Write(common.Uint16ToBytes(b.childrenBitmap()))
+	buffer.Write(encoding)
+	buffer.Write(nibblesToKeyLE(b.key))
+	buffer.Write(common.Uint16ToBytes(b.childrenBitmap()))
 
 	if b.value != nil {
 		bytes, err := scale.Marshal(b.value)
 		if err != nil {
-			return nil, err
-		}
-		h.tmp.Write(bytes)
-	}
-
-	if h.parallel {
-		wg, _ := errgroup.WithContext(context.Background())
-		resBuff := make([][]byte, 16)
-		for i := 0; i < 16; i++ {
-			func(i int) {
-				wg.Go(func() error {
-					child := b.children[i]
-					if child == nil {
-						return nil
-					}
-
-					var err error
-					resBuff[i], err = encodeAndHash(child)
-					if err != nil {
-						return err
-					}
-					return nil
-				})
-			}(i)
-		}
-		if err := wg.Wait(); err != nil {
-			return nil, err
+			return err
 		}
+		buffer.Write(bytes)
+	}
+
+	if parallel {
+		return encodeChildsInParallel(b.children, buffer)
+	}
+	return encodeChildsSequentially(b.children, buffer)
+}
 
-		for _, v := range resBuff {
-			if v != nil {
-				h.tmp.Write(v)
+func encodeChildsInParallel(children [16]node, buffer *bytes.Buffer) (err error) {
```

Comment:
```suggestion
func encodeChildrenInParallel(children [16]node, buffer *bytes.Buffer) (err error) {
```

Reply:
done in 95abb513949f26a6bcebcae4972e5dcd66ad53d0